File: Books/Books/Neural_Networks/python/simple_nn.py
# import libraries
#-------------------------------------------------------#
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
from sklearn.datasets import load_digits
import matplotlib.pyplot as plt 
import numpy.random as r
import numpy as np


# function definitions
#-------------------------------------------------------#   
def f(x):
    # sigmoid(x)
    return 1 / (1 + np.exp(-x))
    
    # tanh(x)
    
    # leakyReLU(x)


def f_deriv(x):
    # d/dx sigmoid(X)
    return f(x) * (1 - f(x))
    
    # d/dx tanh(x)

    # d/dx leakyReLU(x)
    

def feed_forward(x, W, b):
    h = {1: x}
    z = {}
    #
    for l in range(1, len(W) + 1):
        # if it is the first layer, then the input into the weights is x, otherwise, 
        # it is the output from the last layer
        if l == 1:
            node_in = x
        else:
            node_in = h[l]
        #
        z[l+1] = W[l].dot(node_in) + b[l] # z^(l+1) = W^(l)*h^(l) + b^(l)  
        h[l+1] = f(z[l+1]) # h^(l) = f(z^(l)) 
    #    
    return h, z

def setup_and_init_weights(nn_structure):
    W = {}
    b = {}
    #
    for l in range(1, len(nn_structure)):
        W[l] = r.random_sample((nn_structure[l], nn_structure[l-1]))
        b[l] = r.random_sample((nn_structure[l],))
    #    
    return W, b

def init_tri_values(nn_structure):
    tri_W = {}
    tri_b = {}
    #
    for l in range(1, len(nn_structure)):
        tri_W[l] = np.zeros((nn_structure[l], nn_structure[l-1]))
        tri_b[l] = np.zeros((nn_structure[l],))
    #    
    return tri_W, tri_b

def calculate_out_layer_delta(y, h_out, z_out):
    # delta^(nl) = -(y_i - h_i^(nl)) * f'(z_i^(nl))
    return -(y-h_out) * f_deriv(z_out)

def calculate_hidden_delta(delta_plus_1, w_l, z_l):
    # delta^(l) = (transpose(W^(l)) * delta^(l+1)) * f'(z^(l))
    return np.dot(np.transpose(w_l), delta_plus_1) * f_deriv(z_l)

def convert_y_to_vect(y):
    y_vect = np.zeros((len(y), 10))
    for i in range(len(y)):
        y_vect[i, y[i]] = 1
    return y_vect

def train_nn(nn_structure, X, y, iter_num=3000, alpha=0.25):
    W, b = setup_and_init_weights(nn_structure)
    cnt = 0
    m = len(y)
    avg_cost_func = []
    #
    print('Starting gradient descent for {} iterations'.format(iter_num))
    #
    while cnt < iter_num:
        if cnt%1000 == 0:
            print('Iteration {} of {}'.format(cnt, iter_num))
        #    
        tri_W, tri_b = init_tri_values(nn_structure)
        avg_cost = 0
        #
        for i in range(len(y)):
            delta = {}
            # perform the feed forward pass and return the stored h and z values, to be used in the gradient descent step
            h, z = feed_forward(X[i, :], W, b)
            #
            # loop from nl-1 to 1 backpropagating the errors
            for l in range(len(nn_structure), 0, -1):
                if l == len(nn_structure):
                    delta[l] = calculate_out_layer_delta(y[i,:], h[l], z[l])
                    avg_cost += np.linalg.norm((y[i,:]-h[l]))
                else:
                    if l > 1:
                        delta[l] = calculate_hidden_delta(delta[l+1], W[l], z[l])
                    #
                    # triW^(l) = triW^(l) + delta^(l+1) * transpose(h^(l))
                    tri_W[l] += np.dot(delta[l+1][:,np.newaxis], np.transpose(h[l][:,np.newaxis]))
                    #
                    # trib^(l) = trib^(l) + delta^(l+1)
                    tri_b[l] += delta[l+1]
        #
        # perform the gradient descent step for the weights in each layer
        for l in range(len(nn_structure) - 1, 0, -1):
            W[l] += -alpha * (1.0/m * tri_W[l])
            b[l] += -alpha * (1.0/m * tri_b[l])
        #
        # complete the average cost calculation
        avg_cost = 1.0/m * avg_cost
        avg_cost_func.append(avg_cost)
        cnt += 1
        #
    return W, b, avg_cost_func

def predict_y(W, b, X, n_layers):
    m = X.shape[0]
    y = np.zeros((m,))
    for i in range(m):
        h, z = feed_forward(X[i, :], W, b)
        y[i] = np.argmax(h[n_layers])
    return y
    

# start work
#-------------------------------------------------------#    
if __name__ == "__main__":
    # load the MNIST dataset
    digits = load_digits()
    print(digits.data.shape)
    
    # plot the digit that we will test
    plt.figure()
    plt.gray() 
    plt.matshow(digits.images[1]) 
    #plt.savefig('nn_digit.png')
    
    # show the original pixel data
    print('dataset pixel representations \n', digits.data[0,:], '\n\n')    
        
    # show the scaled pixel data
    X_scale = StandardScaler()
    X = X_scale.fit_transform(digits.data)
    print('scale the input data \n', X[0,:], '\n\n')    
    
    # set the digits data    
    y = digits.target
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4)
    
    # convert data to vector
    y_v_train = convert_y_to_vect(y_train)
    y_v_test = convert_y_to_vect(y_test)
    print('y_train[0]=',y_train[0])
    print('y_v_train[0]=', y_v_train[0])
    
    # set the neural net structure
    #     64 nodes to cover the 64 pixels in the image 
    #     30 hidden layer nodes to allow for the complexity of the task
    #     10 output layer nodes to predict the digits
    nn_structure = [64, 30, 10]
    
    # train the neural network
    W, b, avg_cost_func = train_nn(nn_structure, X_train, y_v_train)
    
    # plot the results
    plt.figure()
    plt.plot(avg_cost_func)
    plt.ylabel('Average J')
    plt.xlabel('Iteration number')
    plt.grid(True)
    plt.show()
    #plt.savefig('nn_average_cost_vs_iteration.png')
    
    # show accuracy
    y_pred = predict_y(W, b, X_test, 3)
    print('accuracy score =', accuracy_score(y_test, y_pred)*100)

File: Books/Books/Neural_Networks/matlab/simple_nn.m
% File: simple_nn.m % A MATLAB implementation of a simple neural network 
% adapted from the provided Python code.

%% Main script

% Load dataset 
% (This example assumes you have a MAT file 'digits.mat' containing: 
% data - an [n_samples x 64] matrix of pixel values, 
% images - an [8 x 8 x n_samples] array of images, 
% target - an [n_samples x 1] vector with labels 0-9.) 
load('digits.mat');

% Display dataset dimensions 
disp(size(data));

% Display one digit image (show the second image, since Python used index 1) 
figure; 
colormap(gray); 
imagesc(images(:,:,2)); 
title('Sample Digit');
save_all_figs_OPTION('vnn1','png',1)

% Show raw pixel representation of the first sample 
disp('Dataset pixel representations:'); 
disp(data(1,:));

% Scale the data using zscore (similar to StandardScaler) 
X = zscore(data); 
disp('Scaled input data:'); 
disp(X(1,:));

% Set targets 
y = target;

% Split data into training (60%) and testing (40%) sets 
cv = cvpartition(size(X,1),'HoldOut',0.4); 
X_train = X(training(cv),:); X_test = X(test(cv),:); y_train = y(training(cv)); y_test = y(test(cv));

% Convert labels to one-hot encoding (remember to add 1 for MATLAB indices) 
y_v_train = convert_y_to_vect(y_train); 
y_v_test = convert_y_to_vect(y_test);

disp(['y_train(1)= ', num2str(y_train(1))]); 
disp('y_v_train(1)= '); 
disp(y_v_train(1,:));

% Define neural network structure: 
% 64 input nodes, one hidden layer with 20 nodes, and 10 output nodes. 
nn_structure = [64, 20, 10]; 
iter_num = 5000; 
alpha = 0.5;

% Train the neural network 
[W, b, avg_cost_func] = train_nn(nn_structure, X_train, y_v_train, iter_num, alpha);

% Plot the average cost versus iterations 
figure; 
plot(avg_cost_func); 
ylabel('Average Cost (J)'); 
xlabel('Iteration number'); grid on;

save_all_figs_OPTION('vnn2','png',1)

% Predict on the test set. % Note: use n_layers = length(nn_structure) 
y_pred = predict_y(W, b, X_test, length(nn_structure));

% Calculate accuracy 
accuracy = sum(y_test == y_pred) / length(y_test) * 100; 
disp(['Accuracy score = ', num2str(accuracy)]);

%% Function Definitions
function y = f(x) 
% Sigmoid activation function 
y = 1 ./ (1 + exp(-x)); 
end

function y = f_deriv(x) 
% Derivative of the sigmoid function 
fx = f(x); 
y = fx .* (1 - fx); 
end

function [h, z] = feed_forward(x, W, b) 
% Performs feed-forward propagation. 
% x - input column vector. 
% W - cell array of weight matrices. 
% b - cell array of bias vectors. 
% h - cell array of activations (h{1} is input, h{end} is output). 
% z - cell array of linear combinations.

L = length(W) + 1;  % total number of layers (input + layers with weights)
h = cell(L, 1);
z = cell(L, 1);
h{1} = x;
for l = 1:(L-1)
    node_in = h{l};
    z{l+1} = W{l} * node_in + b{l};
    h{l+1} = f(z{l+1});
end
end

function [W, b] = setup_and_init_weights(nn_structure) 
% Initializes weights and biases randomly. 
% nn_structure is a vector [n_input, n_hidden, n_output]. 
L = length(nn_structure); 
W = cell(L-1, 1); 
b = cell(L-1, 1); 
for l = 1:(L-1) 
	% For layer l, weights are of size (nn_structure(l+1) x nn_structure(l)) 
	W{l} = rand(nn_structure(l+1), nn_structure(l)); 
	b{l} = rand(nn_structure(l+1), 1); 
end 
end

function [tri_W, tri_b] = init_tri_values(nn_structure) 
% Initializes the gradient accumulators (tri_W and tri_b) as zeros. 
L = length(nn_structure); 
tri_W = cell(L-1, 1); 
tri_b = cell(L-1, 1); 
for l = 1:(L-1) 
	tri_W{l} = zeros(nn_structure(l+1), nn_structure(l)); 
	tri_b{l} = zeros(nn_structure(l+1), 1); 
end 
end

function delta = calculate_out_layer_delta(y_true, h_out, z_out) 
% Calculates the output layer delta. 
% y_true and h_out are column vectors. 
delta = -(y_true - h_out) .* f_deriv(z_out); 
end

function delta = calculate_hidden_delta(delta_next, w_l, z_l) 
% Calculates the hidden layer delta. 
delta = (w_l' * delta_next) .* f_deriv(z_l); 
end

function y_vect = convert_y_to_vect(y) 
% Converts label vector y into a one-hot encoded matrix. 
% Assumes y contains digits 0-9. 
n = length(y); 
y_vect = zeros(n, 10); 
for i = 1:n 
% MATLAB indices start at 1, so add 1. 
y_vect(i, y(i) + 1) = 1; 
end 
end

function [W, b, avg_cost_func] = train_nn(nn_structure, X, y, iter_num, alpha) 
% Trains the neural network using gradient descent. 
% nn_structure: network structure vector. 
% X: training data (each row is an example). 
% y: one-hot encoded target matrix. 
% iter_num: number of iterations. 
% alpha: learning rate.

[W, b] = setup_and_init_weights(nn_structure);
m = size(X, 1);
avg_cost_func = zeros(iter_num, 1);
fprintf('Starting gradient descent for %d iterations\n', iter_num);

L = length(nn_structure);  % total number of layers (input + weight layers)

for cnt = 0:(iter_num-1)
    if mod(cnt, 1000) == 0
        fprintf('Iteration %d of %d\n', cnt, iter_num);
    end
    
    [tri_W, tri_b] = init_tri_values(nn_structure);
    avg_cost = 0;
    
    % Loop over all training examples
    for i = 1:m
        x_i = X(i, :)';      % convert row to column vector
        y_i = y(i, :)';      % one-hot target as column vector
        [h, z] = feed_forward(x_i, W, b);
        delta = cell(L, 1);
        
        % Backpropagation (looping from output layer down to input layer)
        for l = L:-1:1
            if l == L
                % Output layer delta
                delta{l} = calculate_out_layer_delta(y_i, h{l}, z{l});
                avg_cost = avg_cost + norm(y_i - h{l});
            else
                if l > 1
                    delta{l} = calculate_hidden_delta(delta{l+1}, W{l}, z{l});
                end
                % Accumulate gradients (outer product of delta and activation)
                tri_W{l} = tri_W{l} + delta{l+1} * h{l}';
                tri_b{l} = tri_b{l} + delta{l+1};
            end
        end
    end
    
    % Update weights and biases using the averaged gradients
    for l = 1:(L-1)
        W{l} = W{l} - alpha * (1/m) * tri_W{l};
        b{l} = b{l} - alpha * (1/m) * tri_b{l};
    end
    
    avg_cost_func(cnt + 1) = avg_cost / m;
end
end

function y_pred = predict_y(W, b, X, n_layers) 
% Predicts labels for the data in X. 
% n_layers should equal length(nn_structure) (input+output layers). 
m = size(X, 1); 
y_pred = zeros(m, 1);

for i = 1:m
    x_i = X(i, :)';
    [h, ~] = feed_forward(x_i, W, b);
    [~, idx] = max(h{n_layers});
    % Subtract 1 to convert from MATLAB’s 1-indexing to the original 0–9 labels.
    y_pred(i) = idx - 1;
end

end
